### PR TITLE
XDP tutorial idea that leads to HW-offloading parsing

### PR DESCRIPTION
**Pull request by netoptimizer**
Hi @qmonnet 

What do you think of this idea, to demonstrate howto use XDP hardware offloading?
--Jesper

### CODE REVIEW

File: areas/xdp-tutorial.org [hunk 1]
```diff
@@ -35,6 +35,18 @@ The [[https://github.com/xdp-project/xdp-tutorial/tree/master/advanced03-AF_XDP]
...
+This exercise/lesson is also meant to show how XDP hardware offloading can
+interact with regular XDP programs (software mode running driver level).
+
+The assignment is to create two XDP programs. The first XDP-prog does packet
+header parsing, and store parse information in metadata area (located just
+before payload) and BPF tail-call second XDP-prog. The second XDP-prog reads
+info from metadata area, and act on parse-info, e.g. show it can read part
+of the packet, without doing parsing. The tricky part of the second XDP-prog
+is how can the verifiers static analysis proof safety.
+
```
Comment (qmonnet) [inline areas/xdp-tutorial.org]: Last line could be improved, maybe something along:

“is how to ensure the verifier is able to perform its static analysis and prove the program is safe”?

File: areas/xdp-tutorial.org [hunk 2]
```diff
@@ -47,6 +47,18 @@ info from metadata area, and act on parse-info, e.g. show it can read part
 of the packet, without doing parsing. The tricky part of the second XDP-prog
 is how can the verifiers static analysis proof safety.
 
+
+*** TODO xdp-tutorial: Extend to HW-offload: Split packet parsing between two XDP programs
+
+When above assignment is complete. Then extend to HW-offload the first
+XDP-prog that does the parsing and storing in metadata area. It doesn't BPF
+tail-call the second XDP-prog, instead load the second XDP-prog in regular
```
Comment (qmonnet) [inline areas/xdp-tutorial.org]: s/load/it loads/

File: areas/xdp-tutorial.org [hunk 3]
```diff
@@ -47,6 +47,18 @@ info from metadata area, and act on parse-info, e.g. show it can read part
...
+XDP driver hook.
+
+This will demonstrate howto HW-offload packet header parse information via
```
Comment (qmonnet) [inline areas/xdp-tutorial.org]: s/howto/how to/

File: areas/xdp-tutorial.org [hunk 4]
```diff
@@ -58,6 +60,8 @@ XDP driver hook.
 This will demonstrate howto HW-offload packet header parse information via
 metadata area.
 
+BTF considerations, how can we transfer BTF info from HW-offloaded XDP-prog?
+
```
Comment (qmonnet) [inline areas/xdp-tutorial.org]: BTF are not working with offloaded programs right now as far as I know. There seems to be a simple fix in the JIT to do, that's on the TODO list but not here yet.

(Also, spurious blank line.)

**Comment by qmonnet** [comment]
Oh and I didn't comment on the idea itself with the two programs working together. It looks good to me, a basic but nice illustration of what can be done by combining several programs.

**Comment by netoptimizer** [comment]
Wow - I didn't expect such a detailed review! Thanks.

**Comment by netoptimizer** [comment]
> Oh and I didn't comment on the idea itself with the two programs working together. It looks good to me, a basic but nice illustration of what can be done by combining several programs.

Thanks for validating that this would be a valuable example.

Any chance that someone from @Netronome could be motivated to code-up this example?

File: areas/xdp-tutorial.org (hunk 1 above)
Comment (netoptimizer) [reply areas/xdp-tutorial.org]: Yes, sounds better

**Comment by qmonnet** [comment]
We don't lack motivation... Just time! :) I'm not sure anyone can work on it right now, but possibly in the future.

**Comment by netoptimizer** [comment]
> We don't lack motivation... Just time! :) I'm not sure anyone can work on it right now, but possibly in the future.

I understand, I have the same problem ;-)

I'll merge this TODO, and then we will see when we have time to cycle back... and if we are lucky, someone (maybe a XDP-newbie) will pick it up and complete the example...

**Pull request merged**